File: ios/Relay/Throttle.swift
import Foundation
final class Throttle {
  private let rate: Int64
  private let burst: Int64
  private let cap: Int64
  private var tokens: Int64
  private var last: TimeInterval
  private var monthUsed: Int64
  init(rateBytesPerSec: Int64, burstBytes: Int64, monthlyCapBytes: Int64){
    self.rate = rateBytesPerSec; self.burst = burstBytes; self.cap = monthlyCapBytes
    self.tokens = burstBytes; self.last = Date().timeIntervalSince1970
    self.monthUsed = UserDefaults.standard.object(forKey:"mon") as? Int64 ?? 0
  }
  func allow(size: Int64) -> Bool {
    let now = Date().timeIntervalSince1970
    let refill = Int64((now - last) * Double(rate))
    tokens = min(burst, tokens + max(0,refill))
    last = now
    if monthUsed + size > cap { return false }
    if tokens < size { return false }
    tokens -= size; monthUsed += size
    UserDefaults.standard.setValue(monthUsed, forKey:"mon")
    return true
  }
}

File: android/app/src/main/java/com/example/relay/Throttle.kt
package com.example.relay
import kotlin.math.min
import java.util.concurrent.atomic.AtomicLong
class Throttle(
  private val rateBytesPerSec: Long,
  private val burstBytes: Long,
  private val monthlyCapBytes: Long,
  private val persist: (Long, Long)->Unit,
  private val load: ()->Pair<Long,Long>
){
  private val lastRef = AtomicLong(System.currentTimeMillis())
  private var tokens = burstBytes
  private var monthUsed = 0L
  init { val (t,m)=load(); tokens=t; monthUsed=m }
  @Synchronized fun allow(size: Long): Boolean {
    val now = System.currentTimeMillis()
    val dt = (now - lastRef.get()).coerceAtLeast(0)
    val refill = (rateBytesPerSec * dt) / 1000
    tokens = min(burstBytes, tokens + refill)
    lastRef.set(now)
    if (monthUsed + size > monthlyCapBytes) return false
    if (tokens < size) return false
    tokens -= size
    monthUsed += size
    persist(tokens, monthUsed)
    return true
  }
  @Synchronized fun resetMonth() { monthUsed = 0; persist(tokens, monthUsed) }
}
